File: api/ocr.py
from flask import Flask, request, jsonify
from flask_cors import CORS
import json
import base64
import io
import re
from PIL import Image
import numpy as np

# Note: PaddleOCR import will be handled with try-catch for deployment
try:
    import paddleocr
    import cv2
    # 日本語特化のPaddleOCR初期化
    ocr = paddleocr.PaddleOCR(
        use_angle_cls=True, 
        lang='japan',  # 日本語認識
        show_log=False
    )
    PADDLE_AVAILABLE = True
except ImportError:
    PADDLE_AVAILABLE = False
    print("PaddleOCR not available, using mock response")

app = Flask(__name__)
CORS(app)  # CORS有効化

@app.route('/api/ocr', methods=['POST', 'OPTIONS'])
def ocr_endpoint():
    if request.method == 'OPTIONS':
        # CORS preflight request
        response = jsonify({'status': 'ok'})
        response.headers.add('Access-Control-Allow-Origin', '*')
        response.headers.add('Access-Control-Allow-Methods', 'POST, OPTIONS')
        response.headers.add('Access-Control-Allow-Headers', 'Content-Type')
        return response

    try:
        # リクエストデータを取得
        data = request.get_json()
        if not data or 'image' not in data:
            return jsonify({
                'success': False,
                'error': 'No image data provided'
            }), 400

        # Base64画像をデコード
        image_data = data['image']
        if image_data.startswith('data:image'):
            image_data = image_data.split(',')[1]
        
        image_bytes = base64.b64decode(image_data)
        image = Image.open(io.BytesIO(image_bytes))
        
        if PADDLE_AVAILABLE:
            # 画像の前処理（車番認識用）
            processed_image = preprocess_license_plate(image)
            
            # PaddleOCRで文字認識
            result = ocr.ocr(processed_image, cls=True)
            detected_text = extract_text_from_paddle_result(result)
        else:
            # 開発用のモックレスポンス（デプロイ時はPaddleOCRが利用可能になる）
            detected_text = "品川 500 あ 12-34"
        
        print(f"検出されたテキスト: {detected_text}")
        
        # 日本のナンバープレート形式をパース
        plate_info = parse_japanese_license_plate(detected_text)
        
        response = {
            'success': True,
            'detected_text': detected_text,
            'plate_info': plate_info,
            'confidence': 95 if PADDLE_AVAILABLE else 85
        }
        
        return jsonify(response)
        
    except Exception as e:
        print(f"エラー: {str(e)}")
        error_response = {
            'success': False,
            'error': str(e),
            'detected_text': '',
            'confidence': 0
        }
        return jsonify(error_response), 500

def preprocess_license_plate(image):
    """車番認識用の画像前処理"""
    # PILからnumpy配列に変換
    img_array = np.array(image)
    if len(img_array.shape) == 3 and img_array.shape[2] == 4:
        # RGBA → RGB
        img_array = img_array[:, :, :3]
    
    if not PADDLE_AVAILABLE:
        return img_array
        
    # OpenCV使用可能な場合の高度な前処理
    try:
        # グレースケール変換
        gray = cv2.cvtColor(img_array, cv2.COLOR_RGB2GRAY)
        
        # コントラスト強化
        clahe = cv2.createCLAHE(clipLimit=3.0, tileGridSize=(8,8))
        enhanced = clahe.apply(gray)
        
        # ガウシアンブラー（ノイズ除去）
        blurred = cv2.GaussianBlur(enhanced, (3, 3), 0)
        
        # 適応的二値化
        binary = cv2.adaptiveThreshold(
            blurred, 255, cv2.ADAPTIVE_THRESH_GAUSSIAN_C, 
            cv2.THRESH_BINARY, 15, 10
        )
        
        return binary
    except:
        return img_array

def extract_text_from_paddle_result(result):
    """PaddleOCRの結果からテキストを抽出"""
    if not result or not result[0]:
        return ""
    
    texts = []
    for line in result[0]:
        if len(line) >= 2:
            text = line[1][0]  # 認識されたテキスト
            confidence = line[1][1]  # 信頼度
            if confidence > 0.5:  # 信頼度50%以上のみ採用
                texts.append(text)
    
    return ' '.join(texts)

def parse_japanese_license_plate(text):
    """日本のナンバープレート形式をパース"""
    print(f"パース対象テキスト: {text}")
    
    # テキストクリーンアップ
    clean_text = text.replace('\n', ' ').replace('\r', ' ')
    clean_text = ' '.join(clean_text.split())  # 余分なスペースを削除
    
    # 日本の地域名リスト（一部）
    regions = [
        '品川', '新宿', '練馬', '世田谷', '杉並', '江東', '足立', '葛飾', '江戸川',
        '板橋', '台東', '墨田', '荒川', '北', '豊島', '中野', '目黒', '大田', '港',
        '千代田', '中央', '文京', '渋谷', '横浜', '川崎', '相模', '湘南', '名古屋',
        '大阪', '神戸', '京都', '福岡', '札幌', '仙台', '広島'
    ]
    
    # ひらがな一覧
    hiragana_list = 'あいうえおかきくけこさしすせそたちつてとなにぬねのはひふへほまみむめもやゆよらりるれろわをん'
    
    result = {
        'region': '',
        'classification': '',
        'hiragana': '',
        'number': '',
        'full_text': clean_text
    }
    
    # パターン1: 完全な形式 "品川 500 あ 12-34"
    pattern1 = r'([^\d\s]{1,4})\s*(\d{3})\s*([あ-ん])\s*(\d{1,2}[-−]\d{2})'
    match1 = re.search(pattern1, clean_text)
    if match1:
        result['region'] = match1.group(1)
        result['classification'] = match1.group(2)
        result['hiragana'] = match1.group(3)
        result['number'] = match1.group(4)
        return result
    
    # パターン2: ハイフンなし "品川 500 あ 1234"
    pattern2 = r'([^\d\s]{1,4})\s*(\d{3})\s*([あ-ん])\s*(\d{4})'
    match2 = re.search(pattern2, clean_text)
    if match2:
        result['region'] = match2.group(1)
        result['classification'] = match2.group(2)
        result['hiragana'] = match2.group(3)
        number = match2.group(4)
        result['number'] = f"{number[:2]}-{number[2:]}"  # 12-34形式に変換
        return result
    
    # パターン3: 部分的にマッチ
    # 地域名を検索
    for region in regions:
        if region in clean_text:
            result['region'] = region
            break
    
    # ひらがなを検索
    hiragana_match = re.search(f'([{hiragana_list}])', clean_text)
    if hiragana_match:
        result['hiragana'] = hiragana_match.group(1)
    
    # 数字パターンを検索
    number_patterns = [
        r'(\d{1,2}[-−]\d{2})',  # 12-34形式
        r'(\d{4})',             # 1234形式
        r'(\d{3})',             # 500形式（分類番号）
    ]
    
    for pattern in number_patterns:
        match = re.search(pattern, clean_text)
        if match:
            num = match.group(1)
            if len(num) == 4 and '-' not in num:
                result['number'] = f"{num[:2]}-{num[2:]}"
            elif len(num) == 3:
                result['classification'] = num
            else:
                result['number'] = num
            break
    
    return result

# Vercel用のハンドラー
def handler(event, context):
    return app(event, context)

File: netlify/functions/ocr.js
// Netlify Functions用のOCR API
// PaddleOCRの代わりにTesseract.jsのNode.js版を使用

const Tesseract = require('tesseract.js');

exports.handler = async (event, context) => {
  // CORS headers
  const headers = {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Headers': 'Content-Type',
    'Access-Control-Allow-Methods': 'POST, OPTIONS'
  };

  if (event.httpMethod === 'OPTIONS') {
    return {
      statusCode: 200,
      headers,
      body: JSON.stringify({ message: 'CORS preflight' })
    };
  }

  if (event.httpMethod !== 'POST') {
    return {
      statusCode: 405,
      headers,
      body: JSON.stringify({ error: 'Method not allowed' })
    };
  }

  try {
    const { image } = JSON.parse(event.body);
    
    if (!image) {
      return {
        statusCode: 400,
        headers,
        body: JSON.stringify({
          success: false,
          error: 'No image provided'
        })
      };
    }

    console.log('OCR処理開始...');

    // Tesseract.jsで日本語＋英語認識（Node.js版は高性能）
    const result = await Tesseract.recognize(image, 'jpn+eng', {
      logger: m => console.log(m),
      // より高精度な設定
      tessedit_pageseg_mode: Tesseract.PSM.SINGLE_WORD,
      tessedit_char_whitelist: '品川新宿練馬世田谷杉並江東足立葛飾江戸川板橋台東墨田荒川北豊島中野目黒大田港千代田中央文京渋谷横浜川崎相模湘南名古屋大阪神戸京都福岡札幌仙台広島あいうえおかきくけこさしすせそたちつてとなにぬねのはひふへほまみむめもやゆよらりるれろわをん0123456789-'
    });

    const detectedText = result.data.text.trim();
    console.log('検出テキスト:', detectedText);

    // 日本のナンバープレートをパース
    const plateInfo = parseJapanesePlate(detectedText);

    return {
      statusCode: 200,
      headers,
      body: JSON.stringify({
        success: true,
        detected_text: detectedText,
        plate_info: plateInfo,
        confidence: Math.round(result.data.confidence)
      })
    };

  } catch (error) {
    console.error('OCRエラー:', error);
    return {
      statusCode: 500,
      headers,
      body: JSON.stringify({
        success: false,
        error: error.message,
        detected_text: '',
        confidence: 0
      })
    };
  }
};

function parseJapanesePlate(text) {
  console.log('パース対象:', text);
  
  const cleanText = text.replace(/\r?\n/g, ' ').replace(/\s+/g, ' ').trim();
  
  const result = {
    region: '',
    classification: '',
    hiragana: '',
    number: '',
    fullText: cleanText
  };
  
  // パターンマッチング
  const patterns = [
    /([^\d\s]{1,4})\s*(\d{3})\s*([あ-ん])\s*(\d{1,2}[-－−]\d{2})/,
    /([^\d\s]{1,4})\s*(\d{3})\s*([あ-ん])\s*(\d{4})/,
    /([^\d\s]{1,4})\s*([あ-ん])\s*(\d{1,2}[-－−]\d{2})/
  ];
  
  for (let i = 0; i < patterns.length; i++) {
    const match = cleanText.match(patterns[i]);
    if (match) {
      if (i === 0) {
        result.region = match[1];
        result.classification = match[2];
        result.hiragana = match[3];
        result.number = match[4];
      } else if (i === 1) {
        result.region = match[1];
        result.classification = match[2];
        result.hiragana = match[3];
        result.number = `${match[4].slice(0, 2)}-${match[4].slice(2)}`;
      } else if (i === 2) {
        result.region = match[1];
        result.hiragana = match[2];
        result.number = match[3];
      }
      console.log('パース成功:', result);
      return result;
    }
  }
  
  // 部分認識
  const regions = ['品川', '新宿', '練馬', '世田谷', '杉並', '江東', '足立'];
  for (const region of regions) {
    if (cleanText.includes(region)) {
      result.region = region;
      break;
    }
  }
  
  // 数字とひらがなの部分検索
  const numberMatch = cleanText.match(/(\d{1,2}[-－−]\d{2}|\d{4})/);
  if (numberMatch) {
    const num = numberMatch[1];
    result.number = num.length === 4 ? `${num.slice(0, 2)}-${num.slice(2)}` : num;
  }
  
  const hiraganaMatch = cleanText.match(/([あ-ん])/);
  if (hiraganaMatch) {
    result.hiragana = hiraganaMatch[1];
  }
  
  return result;
}